src/gpu/intel/jit/gemm/generator/pieces/copy_plan.cpp

```diff
@@ -778,17 +792,96 @@ void CopyPlan::planTypeConversions()
         } else if (st == Type::ngen_f8_e8m0() && dt == DataType::hf) {
                 planEmulatedFP8E8M0ToHF(i);
         } else if (st != dt && (isFP8(st) || isFP8(dt))) {
-            copyThrough(i, DataType::hf, 1);
+            copyThrough(i, DataType::hf);
             rerun = true;
         } else if (st == dt)
-            i.moveToIntegerPipe();
+            planIntegerSwizzle(i);
     }
 
     mergeChanges();
+    if (hw <= HW::Gen12LP)
+        for (auto &i: insns)
+            if (i.op == Opcode::bfn) {
+                rerun = true; break;
+            }
+
     if (rerun)
         planTypeConversions();
 }
 
+void CopyPlan::planIntegerSwizzle(CopyInstruction &i) {
+    if (getBytes(i.src0.type) == 8) {
+        auto &i0 = i, &i1 = split(i);
+        i0.src0.type = DataType::ud;
+        i0.dst.type = DataType::ud;
+        i0.src0.stride *= 2;
+        i0.dst.stride *= 2;
+        i0.src0.offset *= 2;
+        i0.dst.offset *= 2;
+        i1.src0.type = DataType::ud;
+        i1.dst.type = DataType::ud;
+        i1.src0.stride *= 2;
+        i1.dst.stride *= 2;
+        i1.src0.offset *= 2;
+        i1.dst.offset *= 2;
+        i1.src0.offset += 1;
+        i1.dst.offset += 1;
+    } else
+        i.moveToIntegerPipe();
+}
+
+void CopyPlan::planEmulatedBooleanFunction(CopyInstruction &i) {
+    if (i.op != Opcode::bfn || hw > HW::Gen12LP) return;
+    if (i.ctrl != 0xCA) stub();
+
+    auto dst = i.dst, src0 = i.src0, src1 = i.src1, src2 = i.src2;
+
+    // (s0 & ~s2) | (s1 & s2)
+    // t1 <- s0 & ~s2
+    // t2 <- s1 & s2
+    // d <- t2 | t1
+
+    auto ie = splitMultiple<3>(i);
+    ie[0]->op = Opcode::and_;
+    ie[1]->op = Opcode::and_;
+    ie[2]->op = Opcode::or_;
+    for (auto &ins: ie)
+        ins->src2 = {};
+
+    CopyOperand invSrc2 = src2, t1, t2;
+
+    if (src2.kind == CopyOperand::Immediate)
+        invSrc2.value = (~invSrc2.value) & 0xFFFF;
```

Comment:
I'd suggest adding a check that the immediate is in fact 16 bits, in case we ever use 32-bit bfn later.